lib/kitchen_sink/algorithms.ex

```diff
@@ -67,58 +67,112 @@ defmodule KitchenSink.Algorithms do
       iex> Float.round(result, 6)
       10.311285
   """
+  def binary_search(start, finish, fit, target, strategy \\ :midpoint)
   @spec binary_search(
     binary_search_position, binary_search_position, binary_search_fit_func, any, binary_search_strategy
   ) :: binary_search_result
+  def binary_search(
+    range_start_list,
+    range_finish_list,
+    fit,
+    target,
+    strategy
+  ) when is_function(fit) and is_list(range_start_list) and is_list(range_finish_list) do
+    {start_list, finish_list} = ensure_order(range_start_list, range_finish_list)
+    do_binary_search(start_list, finish_list, fit, target, strategy)
+  end
   def binary_search(range_start, range_finish, fit, target, strategy) when is_function(fit) do
-    do_binary_search(range_start, range_finish, fit, target, strategy)
+    binary_search([range_start], [range_finish], fit, target, strategy)
   end
 
-  def binary_search(range_start, range_finish, fit, target) do
-    binary_search(range_start, range_finish, fit, target, :midpoint)
+  defp ensure_order(same, same) do
+    {same, same}
   end
-
-  defp do_binary_search(start, finish, fit, target, strategy) when finish < start do
-    binary_search(finish, start, fit, target, strategy)
+  defp ensure_order(start_list, finish_list) do
+    [a | start_rest] = start_list
+    [b | finish_rest] = finish_list
+    {start, finish} =
+      if a < b do
+        {[a], [b]}
+      else
+        {[b], [a]}
+      end
+    {final_start, final_finish} = ensure_order(start_rest, finish_rest)
+    {start ++ final_start, finish ++ final_finish}
   end
+
   defp do_binary_search(position, position, fit, target, _), do: ok_or_not_found(position, fit, target)
-  defp do_binary_search(start, finish, fit, target, :midpoint) do
-    mid = binary_search_midpoint(start, finish)
-    case fit.(mid, target) do
-      :ok -> {:ok, mid}
-      :high -> do_binary_search(bounded_increment(mid, finish), finish, fit, target, :midpoint)
-      :low -> do_binary_search(start, bounded_decrement(mid, start), fit, target, :midpoint)
+  defp do_binary_search(start_list, finish_list, fit, target, :midpoint) do
```

Comment:
won't the first function catch all the requests? doesn't elixir give a warning for that?

Reply:
ah, I see, clever. :(